### DISCUSSION TITLE
Query Construction: Adding a planner Optimization Barrier or Eager Loading in a 2nd Query

### DISCUSSION THREAD

**Discussion by JLowe-N** [sqlalchemy/sqlalchemy#11185] [Usage Questions]
I am trying to improve a poorly performing query generated by SQL Alchemy.
I am using PostgreSQL 13, SQL Alchemy 1.3, psycopg2-binary 2.9.7, and Python 3.9.2.

We load the modals, but one particular property seems to extend the total query time from 5 seconds per call to 5 milliseconds per query (called 4 times batched in sets of 500 joining on PK ID).

``` 
# These added options are driving almost all the query time
# I believe loading them in bulk in a 2nd query (like a Materialized CTE for example in PostgreSQL) would improve time greatly
    q = q.options(
        joinedload("owner"),
        selectinload("readable_by"),
    )
```

Is there a way to bulk load these properties with an query planner optimization barrier, or to force a separate 2nd query statement to load in bulk?  The default lazy loading option in this case would trigger an N+1.

**Comment by JLowe-N** [comment]
More specifically, I have ~2000 Resource models that are queried.
They all have their own owner & readable by properties.
I would like to load these properties for all the returned models at once, but I would like to force it into its own query (or at least prevent the database from reordering joins as the DB stats can't reliably calculate a good join order).

```
    readable_by = relationship(
        "Lab",
        secondary="resource_read_permission",
        primaryjoin="ReadPermission.resource_id == Resource.resource_id",
        secondaryjoin="ReadPermission.lab_id == Lab.resource_id",
        innerjoin=True,
    )

    owner = relationship("Resource", foreign_keys=[owner_resource_id], remote_side=[resource_id])
```

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you mean by loading them in a second query. That's exactly what `selectinload` does.
If you don't want to use a join, then use a `selectinload` also for the `owner`

**Comment by JLowe-N** [comment]
Hi CaseIIT,

We discovered that selectinload performance was much worse for this property than both `joinedload('readable_by', innerjoin=False)` and `subqueryload('readable_by')`.

I think subqueryload may be what I was looking for as it splits that relationship load into a 2nd SQL statement (vs embedding the relationship into a single statement).

I suspect that the format of the selectinload may give PostgreSQL more flexibility in restructuring the query during query planning which might be leading us to performance troubles.  Have you heard of any issues where `joinedload` or `subqueryload` may be more performant than `selectinload`?

**Reply by CaselIT** [reply]
that's strange, since the selectinload query is much simpler from a db point of view, it's just a plain select with a where-in condition.

subqueryload in general is legacy in v2, but it will likely be present for a long time still, so if that works better feel free to use it.

Do you have a reproducer by any change? Or an example of the sql generated by selectinload that's slower?

cc @zzzeek

**Comment by JLowe-N** [comment]
As I am relying on pg_stat_statements and pg_stat_activity to capture the queries, I don't have easy access to the full list of related queries.
But I do have the plans before/after the swtich from `selectinload('readable_by')` to `joinedload('readable_by', innerjoin=False)`.

`selectinload`
Triggered 4 times, 5 seconds each (paginated at 500, as expected from the docs)
https://explain.dalibo.com/plan/ceab33a8f3cdg56g

`joinedload`
Triggered 1 time, 5 *milliseconds*
https://explain.dalibo.com/plan/ceab33a8f3cdg56g#plan/node/13

It looks like the resource_read_permission join got pushed into another related query, when `q.all()` is called.

```
# 2000 Items 
q = session.query(Item, Item.qty)
q = q.options(
        joinedload("owner"),
        joinedload("readable_by", innerjoin=False), ### <<<< This is the relationship that causes performance issues with selectinload
)

q = q.options(
            joinedload("item_type").options(
                selectinload("resource_resource_vars").joinedload("variable", innerjoin=False),
                selectinload(ItemTypeDefinition.group_memberships).joinedload(ResourceGroupMember.resource_group),
            )
        )

q = q.options(selectinload("resource_vals"))

if sort:
    q = apply_resource_sorts(q, sort)
else:
    q = apply_resource_sort(q, "resource_id", "desc")

q = apply_item_filters(q, Item, session, **filters)

items = q.all()
```

**Reply by CaselIT** [reply]
It's quite wired. I guess sometimes pg thinks it's clever but it's actually not.

To log the query you can also set echo=True on the create engine
